packages/react-theme-provider/src/ThemeProvider.tsx

```diff
@@ -0,0 +1,55 @@
+import * as React from 'react';
+import cx from 'classnames';
+import { useStylesheet } from '@fluentui/react-stylesheets';
+import { tokensToStyleObject } from './tokensToStyleObject';
+import { ThemeContext } from './ThemeContext';
+import { Theme, ThemePrepared } from './types';
+import { mergeThemes } from './mergeThemes';
+import { useTheme } from './useTheme';
+import * as classes from './ThemeProvider.scss';
+
+/**
+ * Props for the ThemeProvider component.
+ */
+export interface ThemeProviderProps extends React.HTMLAttributes<HTMLDivElement> {
+  /**
+   * Defines the theme provided by the user.
+   */
+  theme?: Theme;
+}
+
+/**
+ * ThemeProvider, used for providing css variables and registering stylesheets.
+ */
+export const ThemeProvider = React.forwardRef<HTMLDivElement, ThemeProviderProps>(
+  (
+    { theme, style, className, ...rest }: React.PropsWithChildren<ThemeProviderProps>,
+    ref: React.Ref<HTMLDivElement>,
+  ) => {
+    // Pull contextual theme.
+    const parentTheme = useTheme();
+
+    // Merge the theme only when parent theme or props theme mutates.
+    const fullTheme = React.useMemo<ThemePrepared>(() => mergeThemes(parentTheme, theme), [parentTheme, theme]);
+
+    // Generate the inline style object only when merged theme mutates.
+    const inlineStyle = React.useMemo<React.CSSProperties>(
+      () => tokensToStyleObject(fullTheme.tokens, undefined, { ...style }),
+      [fullTheme, style],
+    );
+
+    // Register stylesheets as needed.
+    useStylesheet(fullTheme.stylesheets);
```

Comment:
Is this batched and performed in an rAF to avoid forced reflow?

Reply:
Currently, no. I tried doing that with `useLayoutEvent` and saw unstyled markup for a second which seems very very bad.

Would RAF avoid this? I don't want users to see unstyled DOM for a millisecond.

Reply:
Have you profiled or benchmarked to see the cost of style recalculation with the proposed approach? Style recalculations can be costly for complex apps with complex selectors.

Reply:
Note that some browsers are still firing rAF after layout (IE is one of them) yet: https://nolanlawson.com/2018/09/25/accurately-measuring-layout-on-the-web/

Reply:
This PR doesn't really register styles. It uses `useStylesheet`, which does the work in the same way load-themed-styles does, with the exception it does it on initial component render, rather than module load.

And even then, stylesheets designed this way can be externalized into a css file to preload them all in one swipe if needed. There are zero runtime class calculations here.